### DISCUSSION TITLE
What is the right way to encode DigestInfo to implement a signature provider?

### DISCUSSION THREAD

**Discussion by matthauck** [openssl/openssl#22109] [Q&A]
I am implementing a signature provider, and it is unclear to me the right way to encode a DigestInfo so that the correct bits are signed. 

I tried declaring a struct that contains a `X509_ALGOR` and `ASN1_OCTET_STRING` members, and serialized like so:
  
```c
ASN1_SEQUENCE( DigestInfo ) = {
     ASN1_SIMPLE( DigestInfo, algorithm, X509_ALGOR ),
     ASN1_SIMPLE( DigestInfo, digest, ASN1_OCTET_STRING ),
} static_ASN1_SEQUENCE_END( DigestInfo );
```

This produced correct asn.1 output, I believe. But, it does not work when verifying signatures with openssl, which always encodes an explicit null value serializing AlgorithmIdentifier parameters. And because the signed bytes have to match exactly when verifying the signature, the null must be present in the signed value, as far as i can tell. 

So, in my provider's `SIGN` function I need to use `encode_pkcs1`, but there doesn't seem to be a way to do so without following suit in hard-coding asn.1 values. https://github.com/openssl/openssl/blob/fc785a554cc37dfa94710b28ced45b03006f0300/crypto/rsa/rsa_sign.c#L242C12-L243

This seems to be what others are doing as well: https://github.com/latchset/pkcs11-provider/blob/30a1e8e39ab4f054738f28b473fcdfb96b9ab877/src/signature.c#L204-L320

This makes it rather difficult to implement a provider. It would be nice if `encode_pkcs1` was exposed somehow. Is there something that I'm missing here?

Thanks.

**Comment by slontis** [comment]
The code is written this way because the FIPS provider does not contain the ASN1 code.
For consistency this was then also used by the default provider.. But before OpenSSL3.0 it used to just use the ASN1 encoder, so youshould be able to do it..

If you look at X509_ALGOR you will see it has a parameters field, which is what you wanted to set to NULL..

e.g.
````
#include <openssl/asn1t.h>

typedef struct {
    X509_ALGOR algorithm;
    ASN1_OCTET_STRING *digest;
} DigestInfo;

DECLARE_ASN1_FUNCTIONS(DigestInfo)

ASN1_SEQUENCE( DigestInfo ) = {
    ASN1_EMBED( DigestInfo, algorithm, X509_ALGOR ),
    ASN1_SIMPLE( DigestInfo, digest, ASN1_OCTET_STRING )
} ASN1_SEQUENCE_END(DigestInfo)

IMPLEMENT_ASN1_FUNCTIONS(DigestInfo)

static int test_sub(void)
{
    BIO *bio = NULL;
    unsigned char digest[32]= { 0,1,2,3,4,5 };
    unsigned char *out = NULL;
    int outlen;
    DigestInfo d;
    ASN1_OBJECT *asn1 = OBJ_txt2obj("SHA256", 0);
    d.digest = ASN1_OCTET_STRING_new();
    ASN1_OCTET_STRING_set(d.digest, digest, sizeof(digest));
    X509_ALGOR_set0(&d.algorithm, asn1, V_ASN1_NULL, NULL);
    outlen = i2d_DigestInfo(&d, &out);

    bio = BIO_new_file("test.der", "wb");
    BIO_write(bio, out, outlen);
    BIO_free(bio);
    return 1;
}
````
which outputs the following which contains a NULL..
````
> dumpasn1 test.der
  0  49: SEQUENCE {
  2  13:   SEQUENCE {
  4   9:     OBJECT IDENTIFIER sha-256 (2 16 840 1 101 3 4 2 1)
 15   0:     NULL
       :     }
 17  32:   OCTET STRING
       :     00 01 02 03 04 05 00 00 00 00 00 00 00 00 00 00
       :     00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
       :   }
````
Then just check that this matches what you would expect

**Reply by matthauck** [reply]
When I try to run this code on 3.0.10, `i2d_DigestInfo` is returning -1. Printing out the error stack, I see:

```
error:068000DE:asn1 encoding routines::illegal zero content
```

**Reply by matthauck** [reply]
(Note: I had to modify it slightly to use `X509_ALGOR*` and `X509_ALGOR_new()` since `X509_ALGOR` is an incomplete type for me, working from the outside.)

**Reply by slontis** [reply]
I used the embed to mainly illustrate that it could be done both ways..

**Comment by slontis** [comment]
Note that if you read the comments in the file containing encode_pkcs1() you will see that the NULL is mandatory.

**Comment by baentsch** [comment]
> This makes it rather difficult to implement a provider. I

FWIW, I would not say that implementing a provider is difficult -- integrating/re-using existing X509 code without either violating API boundaries (or stealing/copying X509 code) is. I faced (and did) the same when doing [oqsprovider](https://github.com/open-quantum-safe/oqs-provider). The provider concept IMO still is work in progress. Personally I think the biggest improvement to it would be made is if the goal were set to completely abandon "default" provider (and re-implement it via the "external" provider API only). But that would entail _significant_ effort.

**Reply by slontis** [reply]
This seems like a good topic to bring up when the provider related meetings happen...

**Reply by levitte** [reply]
Hmmm, externalizing the default provider would be an interesting exercise.  I suspect it will be easier than you anticipate...  but can be proven wrong.

To be noted is that we do use internal libcrypto APIs in the FIPS provider as well.  As a matter of fact, the FIPS provider is made quite a lot from the same sources as the default providers.  There's nothing stopping us from doing that with an externalized default provider as well.

A possible future could see _all_ the legacy lower level crypto APIs are gone from libcrypto, and their source code has moved entirely to providers/implementations.  However, that wouldn't change much how we'd use them from inside our providers, because that's how they are used today.
